src/storage/src/table/streaming_table/state_table.rs

```diff
@@ -316,28 +317,44 @@ impl<S: StateStore> StateTable<S> {
         self.keyspace.table_id()
     }
 
-    /// get the newest epoch of the state store and panic if the `init_epoch()` has never be called
-    pub fn init_epoch(&mut self, epoch: u64) {
-        match self.epoch {
-            Some(prev_epoch) => {
-                panic!(
-                    "init the state table's epoch twice, table_id: {}, prev_epoch: {}, new_epoch: {}",
-                    self.table_id(),
-                    prev_epoch,
-                    epoch
-                )
-            }
-            None => {
-                self.epoch = Some(epoch);
-            }
+    /// Get the newest epoch of the state store and panic if the `init_epoch()` has never be called.
+    /// The `epoch` should be a previous epoch.
+    pub async fn init_epoch(&mut self, epoch: u64) {
+        if let Some(prev_epoch) = self.epoch {
+            panic!(
+                "init the state table's epoch twice, table_id: {}, prev_epoch: {}, new_epoch: {}",
+                self.table_id(),
+                prev_epoch,
+                epoch
+            )
         }
+
+        self.epoch = Some(epoch);
+        self.wait_epoch().await;
     }
 
     /// get the newest epoch of the state store and panic if the `init_epoch()` has never be called
     pub fn epoch(&self) -> u64 {
         self.epoch.unwrap_or_else(|| panic!("try to use state table's epoch, but the init_epoch() has not been called, table_id: {}", self.table_id()))
     }
 
+    /// Wait for the current epoch of the state store is committed, this is necessary for scaling.
+    ///
+    /// When scaling,
+    /// - some new executors will be created (with `init_epoch`),
+    /// - or the vnode bitmap of existing executors will be changed (with `update_vnode_bitmap`).
+    /// Therefore, the state table might be used to operate on the data from new partitions which is
```

Comment:
other than scaling and vnode change, we should also wait epoch on recovery because shared buffer will be cleared and the cached version can be stale. In general, i think we need to wait epoch on actor creation.

I believe this is already covered in this pr because we wait epoch on state table init epoch.

Reply:
> In general, i think we need to wait epoch on actor creation.

As creating actors is done asynchronously, I think what you mean is "when actors start running" or "actually connected to the graph". I find that too dirty to do this on the state table layer as this requires to handle the calling order of "commit/update epoch", "update vnode mapping/init epoch", `yield barrier` very carefully.

From another perspective, it might be the global barrier manager's responsibility to ensure that versions have been visible to all compute nodes. So I'd like to handle this on the meta service, which can also resolve the same issue when recovers. cc @yezizp2012